### PR TITLE
Gatsby copy linked files, video tag support, a tag support and option to ignore default ignored file types

### PR DESCRIPTION
**Pull request by chiedo**
I added the ability to copy linked images in HTML nodes and also the ability to override which file types are ignored so you can now copy png and svg files if you would like, etc.

### CODE REVIEW

**Comment by chiedo** [comment]
video and a tag support coming soon.

**Comment by chiedo** [comment]
I think this is officially ready. :) All features I wanted to add are now there.

File: packages/gatsby-remark-copy-linked-files/README.md [hunk 1]
```diff
@@ -15,7 +15,12 @@ plugins: [
     resolve: `gatsby-transformer-remark`,
     options: {
       plugins: [
-        `gatsby-remark-copy-linked-files`,
+        { 
+          resolve: 'gatsby-remark-copy-linked-files',
+          options: {
+            ignoreFileExtensions: ['png', 'jpg', 'jpeg', 'bmp', 'tiff'],
```
Comment (KyleAMathews) [inline packages/gatsby-remark-copy-linked-files/README.md]: So you don't need to add it.

Or even better, show the simple default way to add the plugin to gatsby-config and what to do if you want to copy images over (e.g. an empty `ignoreFileExtensions`.

Comment (chiedo) [inline packages/gatsby-remark-copy-linked-files/README.md]: Yes sir!

Comment (chiedo) [inline packages/gatsby-remark-copy-linked-files/README.md]: Let me know what you think about the changes I just made.

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 12a333472c97b480824fa36b3c720c274faf85c0

https://app.netlify.com/sites/using-styled-components/deploys/5977994b6f4c50758d7ec9af

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 12a333472c97b480824fa36b3c720c274faf85c0

https://app.netlify.com/sites/using-remark/deploys/5977994b6f4c50758d7ec9b7

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 12a333472c97b480824fa36b3c720c274faf85c0

https://app.netlify.com/sites/using-glamor/deploys/5977994b6f4c50758d7ec9ac

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 12a333472c97b480824fa36b3c720c274faf85c0

https://app.netlify.com/sites/image-processing/deploys/5977994b6f4c50758d7ec9a8

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 12a333472c97b480824fa36b3c720c274faf85c0

https://app.netlify.com/sites/using-contentful/deploys/5977994b6f4c50758d7ec9b2

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 12a333472c97b480824fa36b3c720c274faf85c0

https://app.netlify.com/sites/using-postcss-sass/deploys/597799496f4c50758d7ec9a1

**Comment by KyleAMathews** [comment]
Getting a bunch of errors now when building gatsbyjs.org

```
3:50:36 PM: UNHANDLED REJECTION TypeError: Expected a string
at module.exports (/opt/build/repo/www/node_modules/is-absolute-url/index.js:4:9)
at module.exports (/opt/build/repo/www/node_modules/is-relative-url/index.js:5:10)
at Node.<anonymous> (/opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:176:21)
at initialize.exports.each (/opt/build/repo/www/node_modules/cheerio/lib/api/traversing.js:300:24)
at _callee2$ (/opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:175:24)
at tryCatch (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:65:40)
at Generator.invoke [as _invoke] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:303:22)
at Generator.prototype.(anonymous function) [as next] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:117:21)
at step (/opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:35:14
at new Promise (/opt/build/repo/www/node_modules/core-js/library/modules/es6.promise.js:191:7)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:14:12
at /opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:454:20
at one (/opt/build/repo/www/node_modules/unist-util-visit/index.js:23:16)
at all (/opt/build/repo/www/node_modules/unist-util-visit/index.js:44:20)
at one (/opt/build/repo/www/node_modules/unist-util-visit/index.js:27:14)
3:50:36 PM: UNHANDLED REJECTION TypeError: Expected a string
at module.exports (/opt/build/repo/www/node_modules/is-absolute-url/index.js:4:9)
at module.exports (/opt/build/repo/www/node_modules/is-relative-url/index.js:5:10)
at Node.<anonymous> (/opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:353:21)
at initialize.exports.each (/opt/build/repo/www/node_modules/cheerio/lib/api/traversing.js:300:24)
at _callee2$ (/opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:352:22)
at tryCatch (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:65:40)
at Generator.invoke [as _invoke] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:303:22)
at Generator.prototype.(anonymous function) [as next] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:117:21)
at step (/opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:35:14
at new Promise (/opt/build/repo/www/node_modules/core-js/library/modules/es6.promise.js:191:7)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:14:12
at /opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:454:20
at one (/opt/build/repo/www/node_modules/unist-util-visit/index.js:23:16)
at all (/opt/build/repo/www/node_modules/unist-util-visit/index.js:44:20)
at one (/opt/build/repo/www/node_modules/unist-util-visit/index.js:27:14)
3:50:46 PM: UNHANDLED REJECTION TypeError: Cannot read property 'slice' of undefined
at _callee3$ (/opt/build/repo/www/node_modules/gatsby-remark-images/index.js:258:51)
at tryCatch (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:65:40)
at Generator.invoke [as _invoke] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:303:22)
at Generator.prototype.(anonymous function) [as next] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:117:21)
at step (/opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:35:14
at new Promise (/opt/build/repo/www/node_modules/core-js/library/modules/es6.promise.js:191:7)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:14:12
at /opt/build/repo/www/node_modules/gatsby-remark-images/index.js:338:26
at Promise._execute (/opt/build/repo/www/node_modules/bluebird/js/release/debuggability.js:300:9)
at Promise._resolveFromExecutor (/opt/build/repo/www/node_modules/bluebird/js/release/promise.js:483:18)
at new Promise (/opt/build/repo/www/node_modules/bluebird/js/release/promise.js:79:10)
at /opt/build/repo/www/node_modules/gatsby-remark-images/index.js:205:16
at Array.map (native)
at /opt/build/repo/www/node_modules/gatsby-remark-images/index.js:204:20
at tryCatcher (/opt/build/repo/www/node_modules/bluebird/js/release/util.js:16:23)
3:50:47 PM: UNHANDLED REJECTION TypeError: Expected a string
at module.exports (/opt/build/repo/www/node_modules/is-absolute-url/index.js:4:9)
at module.exports (/opt/build/repo/www/node_modules/is-relative-url/index.js:5:10)
at Node.<anonymous> (/opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:176:21)
at initialize.exports.each (/opt/build/repo/www/node_modules/cheerio/lib/api/traversing.js:300:24)
at _callee2$ (/opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:175:24)
at tryCatch (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:65:40)
at Generator.invoke [as _invoke] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:303:22)
at Generator.prototype.(anonymous function) [as next] (/opt/build/repo/www/node_modules/regenerator-runtime/runtime.js:117:21)
at step (/opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:35:14
at new Promise (/opt/build/repo/www/node_modules/core-js/library/modules/es6.promise.js:191:7)
at /opt/build/repo/www/node_modules/babel-runtime/helpers/asyncToGenerator.js:14:12
at /opt/build/repo/www/node_modules/gatsby-remark-copy-linked-files/index.js:454:20
at one (/opt/build/repo/www/node_modules/unist-util-visit/index.js:23:16)
at all (/opt/build/repo/www/node_modules/unist-util-visit/index.js:44:20)
at one (/opt/build/repo/www/node_modules/unist-util-visit/index.js:27:14)
3:50:47 PM: UNHANDLED REJECTION TypeError: Expected a string
```

**Comment by chiedo** [comment]
I'll look into it! Worked on all my sites. I'm assuming the deploy preview will tell me when I got it right?

**Comment by chiedo** [comment]
Where can I find the gatsbyjs.org code base so I can do some test builds and debug? Grep and find turned up nothing. Sorry if I'm missing something obvious.

**Comment by KyleAMathews** [comment]
It's in the `www` directory.

**Comment by chiedo** [comment]
I haven't had any luck. If anyone else sees this, feel free to dig in.

This plugin works perfectly on my sites. I can't figure out why it isn't working on gatsby.org.

At this point the only errors I'm getting are gatsby-remark-images errors which doesn't make sense to me seeing as I've changed nothing in those files in this pull request.

It seems like there's some sort of connection between the two plugins even when this plugin ignores the file types that the remark-images plugin works on.

Not sure. I'm using this in production as a local plugin on my sites but can't figure out why this is breaking gatsby.org. Spent about 3 hours digging into it with no progress.

Open to answer any questions for anyone that sees this and has them

**Comment by KyleAMathews** [comment]
If you remove this plugin does Gatsbyjs build? If you change the order of plugins does that change anything?

**Comment by chiedo** [comment]
The issues remain if I comment out this plugin. The exact same error (all gatsby-remark-images errors after adding some try catches) still displays.

The issues go away if I comment out gatsby-remark-images.

That's why I'm confused. I haven't changed anything in remark-images as far as this branch is concerned... lol

It appears to be a gatsby-remark-images bug but I don't understand how that makes sense.

**Comment by KyleAMathews** [comment]
Hmmm that doesn't make sense. Can you reproduce these errors if you switch to master?

**Comment by chiedo** [comment]
First thing I tried this morning. Master worked fine.

Just tested again. Master is still error free.

How are you testing plugins by the way that aren't public yet? I'm manually overwriting node_modules to test how they would function from npm. I'm using the plugins directory fore my projects while waiting for pull requests to be merged but while testing before doing a pull request, is there an easier way than copying to and clearing node_modules?

**Comment by KyleAMathews** [comment]
Update to master then?

On developing, gatsby-dev-cli https://www.gatsbyjs.org/docs/how-to-contribute/

**Comment by chiedo** [comment]
> On developing, gatsby-dev-cli https://www.gatsbyjs.org/docs/how-to-contribute/

Thanks, I'll have to read that section more closely!

> Update to master then?

Same issue as soon as I add my new version of the gatsby-remark-copy-linked-files plugin. As soon as I change that, gatsby-remark-images starts having issues.

**Comment by KyleAMathews** [comment]
Have you tried changing the order of the two plugins? If there's some sort of interaction between the two, changing the order probably would affect that.

**Comment by chiedo** [comment]
Yeah, no difference. I've tried a lot and I'm honestly quite confused. I don't even see how interaction could occur. My code isn't doing anything that should be affected by any interaction. It's ultimately just reading files and then copying data.

**Comment by KyleAMathews** [comment]
Yeah!!! Tests are passing and gatsbyjs.org is building :-D

Merging this overly old PR now!

**Comment by chiedo** [comment]
YAY!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! :)

File: packages/gatsby-remark-copy-linked-files/README.md [hunk 2]
```diff
@@ -42,3 +73,14 @@ it.
...
+### Supported Markdown tags
+
+- img
+- link
+
+### Supported HTML tags
+
+- <img />
+- <video />
+- <a />
```
Comment (jakedeichert) [inline packages/gatsby-remark-copy-linked-files/README.md]: These are not wrapped in code tags `` and are attempting to render on https://www.gatsbyjs.org/packages/gatsby-remark-copy-linked-files/

File: packages/gatsby-remark-copy-linked-files/src/index.js [hunk 3]
```diff
@@ -3,9 +3,19 @@ const isRelativeUrl = require(`is-relative-url`)
 const fsExtra = require(`fs-extra`)
 const path = require(`path`)
 const _ = require(`lodash`)
-const $ = require(`cheerio`)
+const cheerio = require(`cheerio`)
+const sizeOf = require(`image-size`)
```
Comment (jakedeichert) [inline packages/gatsby-remark-copy-linked-files/src/index.js]: This new dependency `image-size` was added yet not included in the package.json which causes this plugin to fail for me

**Comment by jakedeichert** [comment]
Hey guys, found two issues that I just ran into.

I'll submit a PR tonight or tomorrow.

**Pull request merged**